store/src/state.rs

```diff
@@ -102,14 +180,127 @@ impl State {
+    /// - a transaction is open in the DB and the writes are started
+    ///  - while the transaction is not commited, concurrent reads are allowed, both the DB and
+    ///  the in-memory representations, which are consistent at this stage.
+    /// - prior to commiting the changes to the DB, an exclusive lock to the in-memory data is
+    /// acquired, preventing concurrent reads to the in-memory data, since that will be
+    /// out-of-sync w.r.t. the DB.
+    /// - the DB transaction is commited, and requests that read only from the DB can proceed to
+    /// use the fresh data
+    /// - the in-memory structures are updated, and the lock is released
+    pub async fn apply_block(
+        &self,
+        new_block: NewBlock,
+        nullifiers: &[RpoDigest],
+        accounts: &[(AccountId, Digest)],
+        notes: &[Note],
+    ) -> Result<Option<block_header::BlockHeader>, anyhow::Error> {
+        let _ = self.writer.try_lock().map_err(|_| StateError::ConcurrentWrite)?;
+
+        // ensures the right block header is being processed
+        let prev_block_msg = self
+            .db
+            .select_block_header_by_block_num(None)
+            .await?
+            .ok_or(StateError::DbBlockHeaderEmpty)?;
+        let block_num = (prev_block_msg.block_num as u32) + 1;
+        let prev_block: BlockHeader = prev_block_msg.try_into()?;
+        let prev_hash = prev_block.hash();
+        if new_block.prev_hash != prev_hash {
+            return Err(StateError::NewBlockInvalidPrevHash.into());
+        }
+
+        // scope to read in-memory data, validate the request, and compute intermediary values
+        let (account_tree, chain_mmr, nullifier_tree, block_header) = {
+            let inner = self.inner.read().await;
+
+            // nullifiers can be produced only once
+            let duplicate_nullifiers: Vec<_> = nullifiers
+                .iter()
+                .filter(|&&n| inner.nullifier_tree.get_value(n) != EMPTY_WORD)
+                .cloned()
+                .collect();
+            if !duplicate_nullifiers.is_empty() {
+                return Err(StateError::DuplicatedNullifiers(duplicate_nullifiers).into());
+            }
+
+            // update the in-memory datastructures and compute the new block header. Important, the
+            // structures are not yet commited
+            let mut chain_mmr = inner.chain_mmr.clone();
+            chain_mmr.add(prev_hash);
+            let peaks = chain_mmr.peaks(chain_mmr.forest())?;
+
+            let mut nullifier_tree = inner.nullifier_tree.clone();
+            let nullifier_data = block_to_nullifier_data(block_num);
+            for nullifier in nullifiers {
+                nullifier_tree.insert(*nullifier, nullifier_data);
+            }
+
+            let mut account_tree = inner.account_tree.clone();
+            for (account_id, account_hash) in accounts {
+                account_tree.update_leaf(*account_id, account_hash.try_into()?)?;
+            }
```

Comment:
Maybe not for this PR, but it would be cool to add some logging here to understand how much time is spent on updating various in-memory structures, writing to DB etc. Let's create an issue for this.

Reply:
added tracing's span for in-memory updates + transaction.

Note: The measurement of the transaction includes async <=> sync communication. I'm  not sure if it makes sense to measure the sqlite time without including the commit, and the commit is guarded by the oneshot channel.